### ISSUE TITLE
Improperly configured sshd_config file with step ssh config

### ISSUE THREAD

**Issue by TopherIsSwell** [smallstep/cli#232]
### Subject of the issue
```step ssh config --host``` generates an invalid sshd_config file.

### Your environment
* OS - RHEL 7.7 / Centos 7.7 (produced on both)
* Version - Step 0.14.0 RC3

### Steps to reproduce
Set up an step-ca server with ssh support as documentation suggests(e.g.```step ca init --ssh```, etc.)
On a client server (or the same server) generate a signed host key and attempt to install it with step:
```bash
curl -LO https://github.com/smallstep/cli/releases/download/v0.14.0-rc.3/step_0.14.0-rc.3_linux_amd64.tar.gz
tar xf step_0.14.0-rc.3_linux_amd64.tar.gz 
sudo cp step_0.14.0-rc.3/bin/step /usr/local/bin/
# modify as appropriate:
sudo /usr/local/bin/step ca bootstrap --ca-url "https://step-ca-server.domain.com:443" --fingerprint "XXXXXX"
# sign existing host key
sudo /usr/local/bin/step ssh certificate --host --sign $(hostname) /etc/ssh/ssh_host_rsa_key.pub 
# install certificate
sudo /usr/local/bin/step ssh config --host
```

### Expected behaviour
The sshd_config file should be configured to point at the certificate and CA files. /etc/sshd_config should have a valid config. Specifically there should be lines for the following:
```
HostCertificate /etc/ssh/ssh_host_rsa_key-cert.pub
HostKey /etc/ssh/ssh_host_rsa_key
```

### Actual behaviour
I get something akin to the following:
```bash
$ tail -n5 /etc/ssh/sshd_config
# autogenerated by step @ 2020-03-16T23:31:28Z
TrustedUserCAKeys /etc/ssh/ca.pub
HostCertificate /etc/ssh/<no value>
HostKey /etc/ssh/<no value>
# end
```
furthermore, this isn't a valid sshd_config:
```bash
$ sudo sshd -t
/etc/ssh/sshd_config line 56: garbage at end of line; "value>".
```

I haven't tested this, but I assume I would not be able to SSH back into this machine in this configuration following a reboot, which could be a rather large liability. If you're unable to fix this issue by the next release, I'd strongly recommend masking this command to prevent users from munging their configs and locking themselves out of their servers.

**Comment by mmalone** [comment]
Not sure about this particular issue (will let someone else more familiar with that code base chime in) but to address this category of issue perhaps we should check the config before accepting it. Something like:

  1. Copy `sshd_config` to `sshd_config.bak`
  2. Make changes
  3. Run `sshd -t`
  4. If config is good, `rm sshd_config.bak`
  5. Else, `mv sshd_config.bak sshd_config` and print error

People can already do this themselves, but it feels like the consequences of a mistake here justify some caution.

**Comment by maraino** [comment]
The problem here is that the default sshd_config template requires to user suplied variables:

sshd_config.tpl is:
```
TrustedUserCAKeys /etc/ssh/ca.pub
HostCertificate /etc/ssh/{{.User.Certificate}}
HostKey /etc/ssh/{{.User.Key}}
```

Step will be able to fill those gaps only if these options are set, for example:
```
step ssh config --host --set Certificate=ssh_host_ecdsa_key-cert.pub --set Key=ssh_host_ecdsa_key
```

**Comment by maraino** [comment]
So basically the process to configure the ssh server requires at least two commands:

```sh
step ssh certificate --host --sign ssh.hostname.com /etc/ssh/ssh_host_ecdsa_key.pub
step ssh config --host --set Certificate=ssh_host_ecdsa_key-cert.pub --set Key=ssh_host_ecdsa_key
```

The second command will depend on the options you use in the first command, for example, if instead of signing an existent key, a new one is created you will need to set up the name of the new certificate and key.

**Comment by TopherIsSwell** [comment]
Perhaps, then ```step ssh config --host``` should fail lest those variables are set to valid filenames in the /etc/ssh directory.

**Comment by maraino** [comment]
Those variables are template specific, so we cannot fail before going to step-certificates. But we can look into `missingkey=error` option in https://golang.org/pkg/html/template/#Template.Option

We cannot fail before calling step-certificates because the user can use a different template that does not require those variables.

**Comment by maraino** [comment]
I see a few options to manage this type of errors:

1. The one mentioned before, add the option `missingkey=error`. The problem with this is that we won't be able to do useful things like (so I would not recommend this option):
```
{{- if .User.User }}
	User {{.User.User}}
{{- end }}
```
2. Change the templates to fail in case something required is missing, we can manually force an error in the template if some preconditions are not met. Returning the proper error to the cli might be a little bit challenging.
3. For that specific template surround the lines that require user input with if conditions and render a template without those. The problem would be that an ssh server won't be fully configured, only client authentication (`TrustedUserCAKeys`) will be enabled.

@mmalone @TopherIsSwell What do you think?

**Comment by TopherIsSwell** [comment]
Well, applying a little bit of POLA, I'd say that if we can get an error back to the CLI reliably, to take option 2 and fail entirely if the information is missing. If returning a error at the CLI isn't tenable, I'd recommend going with option 3. Because it will at least enable client auth, if not host validation. And since there wouldn't be an error, I think a user would expect some behavior to happen.

Also, 90% of the usefulness of SSH certificates (to me at least) is the client auth portion. The host validation bit is icing on the cake, but not critical to (what I would assume to be) the average workflow. Furthermore, I don't think the ```step ssh config``` for clients sets up a trust for the CA, so the hosts still show as untrusted even if the ```step ssh config --host``` sets up the host certificate correctly. So I'd say getting client auth to work reliably is much more relevant than the host certificate portion.

The more I think about it, I think I'm #teamOption3 even if we can return an error back to the user.

**Comment by mmalone** [comment]
I like being explicit, so my first inclination is to require specification of any template variables. That could become impractical if we start templating larger configuration files with lots of optional variables though. It could also break existing clients when templates change and a variable is added.

I do think that, in this particular case, running `sshd -t` before accepting the rendered configuration is probably the best solution. It pushes the validation off onto OpenSSH, which is a better position to do this work. For example, it'll be easier for OpenSSH to check that the `Certificate` parameter is not just set, but is set to a string that points to a filename that contains a valid certificate.

The problem with just running `sshd -t` is, while I believe it would prevent the existential threat of being locked out of your system in this case, it's not a general solution. And it wouldn't have provided any useful information regarding _why_ the failure occurred.

I don't actually love option three in this case. I'm worried that, instead of failing loud and fast, it will silently proceed with a partially configured server and people will just get confused later. Parenthetically, `step ssh config` actually does configure clients to validate host certificates. You're right that it doesn't go in `~/.ssh/known_hosts` though. Instead, `~/.ssh/config` includes `$(step path)/ssh/config` which specifies `UserKnownHostsFile $(step path)/ssh/known_hosts`. If you look in `$(step path)/ssh/known_hosts` you'll find your host CA public key. It's done this way so that all of the `step`-specific SSH config (except the include directive) is in one place.

Here's a proposal:

1. For this case, run `sshd -t` before accepting a configuration (as described above)
2. Make template variables _required_
3. Add a flag (e.g., `--allow-unset-variables`) that makes template variables _optional_

Without the `--allow-unset-variables` flag the CLI could produce an error that lists any unset variables, which should help people figure out what's going on. I think it'll be pretty easy to implement since the `--allow-unset-variables` flag will map pretty directly to the `missingkey=error` option.

The (rather large) downside is that it would break any existing automation when a template variable is added. One way to fix this would be to have some way to specify whether a variable is required or optional _in the template_ versus doing so at the command line. I'm not sure the template engine we're using supports such a thing though.

Another way to fix this would be to have _conditionals_ not error if a variable is unset, so `{{- if .User.User }}` would work (and an unset `.User.User` would coerce to `false`), but make rendering `{{.User.User}}` fail. Then we could probably get away with not having the `--allow-unset-variables` flag and just saying "all rendered template variables are required", which makes a lot of sense to me. Again, not sure if the template engine supports this.

**Comment by maraino** [comment]
With the fix on #291 step-ca will complain if it detects a variable that is not present. Due to the `text/template` limitations, we cannot safely detect the missing variables without reducing flexibility, so we've decided to be able to configure the required variables in the ca.json. Certificate and Key that are the only ones required in the default templates will be automatically backfilled if the same template is used.
